### PR TITLE
Avoid possible personally identifiable information in analytics

### PR DESCRIPTION
**Pull request by tombogle**
---
This change is [Reviewable](https://reviewable.io/reviews/hatton/desktopanalytics.net/16)

### CODE REVIEW

File: src/DesktopAnalytics/Analytics.cs [hunk 1]
```diff
@@ -63,17 +66,35 @@ public Analytics(string apiSecret, UserInfo userInfo, Dictionary<string, string>
 			_singleton = this;
 			_propertiesThatGoWithEveryEvent = propertiesThatGoWithEveryEvent;
 
-			_userInfo = userInfo;
+			if (retainPii)
```
Comment (hatton) [inline src/DesktopAnalytics/Analytics.cs]: I like all that you're doing here, and I don't see any problems. I am nervous about the string manipulation without unit tests. If there were a bug in all of this manipulation, what is the worst that could happen?

File: src/DesktopAnalytics/Analytics.cs [hunk 2]
```diff
@@ -369,20 +398,32 @@ private void ReportIpAddressOfThisMachineAsync()
...
+								Debug.WriteLine(String.Format("DesktopAnalytics: external ip = {0}", result));
+								_options.Context.Add("ip", result);
+								_propertiesThatGoWithEveryEvent.Add("ip", result);
+							}
 						}
 						catch (Exception)
 						{
 							// we get here when the user isn't online, or anything else prevents us from
-							// getting their ip. Still worth reporting the launch in the later case.
+							// getting their ip. Still worth reporting the launch in the latter case.
```
Comment (hatton) [inline src/DesktopAnalytics/Analytics.cs]: 1) Getting there *location*"

2) I don't understand the original code... if we get the info we want, we add a new piece of info to the properties that go with ever event. Cool. But if we don't get the info, then we call `TrackWith...("Launch")`?

File: src/DesktopAnalytics/Analytics.cs [hunk 3]
```diff
@@ -357,10 +378,18 @@ public static void IdentifyUpdate(UserInfo userInfo)
...
 		/// </summary>
+		/// <remarks>This used to default to "http://icanhazip.com"; //formerly: "http://ipecho.net/plain" (that URL went down, but is now back up)</remarks>
 		public static string UrlThatReturnsExternalIpAddress { get; set; }
+		/// <summary>
+		/// Override this for any reason you like, including if the built-in one ( http://ip-api.com/json/) stops working some day.
+		/// This will be ignored if <seealso cref="UrlThatReturnsExternalIpAddress"/> is set.
+		/// The service should return json that contains values for one or more of the following names: city, country, countryCode,
+		/// region, regionName.
+		/// </summary>
+		public static string UrlThatReturnsGeolocationJson { get; set; }
```
Comment (hatton) [inline src/DesktopAnalytics/Analytics.cs]: Not news to you, but the problem here is that unlike ip, the json codes may well not match. I was looking at our current default here, and SSL is a paid option. It seems likely that HTTP could cause problems in the future. So I found https://ipstack.com/, but it, for example, has "country_code". (it also does not give away SSL, btw).  Ah well.

Comment (hatton) [reply src/DesktopAnalytics/Analytics.cs]: So if we were serious about making his configurable, we could change this to a method that takes a url and then 3 or 4 parameters corresponding to the json names to get at country, city, etc.

File: src/DesktopAnalytics/Analytics.cs (hunk 2 above)
Comment (tombogle) [reply src/DesktopAnalytics/Analytics.cs]: I guess the point is that as long as they are on-line, we can still note that the program was launched (so we can count it) even if we couldn't figure out anything about the location. This is probably a pretty unlikely scenario, but if the URL(s) for getting that info disappear or are blocked by a particular country or ISP, we can at least know the program is being used somewhere.

File: src/DesktopAnalytics/Analytics.cs (hunk 3 above)
Comment (tombogle) [reply src/DesktopAnalytics/Analytics.cs]: Yeah, actually I didn't really think that far ahead. My slightly-more-configurable option was a response to the other URLs keeling over. In reality, if the current solution stops working, we'll probably want to fix it, but by letting the client configure it, they aren't stuck waiting on us, and they might be able to overcome some challenge specific to their needs. Do you think I should go ahead and make it more fully configurable?

Comment (hatton) [reply src/DesktopAnalytics/Analytics.cs]: No, because really, "they" are just you and me :-)

Comment (tombogle) [reply src/DesktopAnalytics/Analytics.cs]: It's on nuget. There could be a million users... :-)

**Comment by tombogle** [comment]
I pushed some unit tests yesterday. Let me know if you see anything else that needs attention. I didn't edit the build configurations to run the tests. (There isn't one for PRs.) But I'm guessing you'll want to do that.

**Pull request merged**